### PR TITLE
`azurerm_container_app_environment` - add support for the Custom DNS Suffix (via the new `azurerm_container_app_environment_custom_domain` resource) and expose the `custom_domain_verification_id` attribute

### PR DESCRIPTION
**Pull request by iambaim**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR:

1. Enables the Custom environment DNS Suffix in Azure Container Apps (https://learn.microsoft.com/en-us/azure/container-apps/environment-custom-dns-suffix) by creating a new virtual resource `azurerm_container_app_environment_custom_domain`.
2. Exposes the Custom Domain Validation ID (`custom_domain_verification_id` attribute) in `azurerm_container_app_environment`'s resource and data source. 

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 

## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [X] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containerapps -parallel=1 -run=TestAccContainerAppEnvironmentCustomDomainResource_ -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccContainerAppEnvironmentCustomDomainResource_basic
=== PAUSE TestAccContainerAppEnvironmentCustomDomainResource_basic
=== RUN   TestAccContainerAppEnvironmentCustomDomainResource_requiresImport
=== PAUSE TestAccContainerAppEnvironmentCustomDomainResource_requiresImport
=== RUN   TestAccContainerAppEnvironmentCustomDomainResource_update
=== PAUSE TestAccContainerAppEnvironmentCustomDomainResource_update
=== CONT  TestAccContainerAppEnvironmentCustomDomainResource_basic
--- PASS: TestAccContainerAppEnvironmentCustomDomainResource_basic (859.72s)
=== CONT  TestAccContainerAppEnvironmentCustomDomainResource_update
--- PASS: TestAccContainerAppEnvironmentCustomDomainResource_update (1025.33s)
=== CONT  TestAccContainerAppEnvironmentCustomDomainResource_requiresImport
--- PASS: TestAccContainerAppEnvironmentCustomDomainResource_requiresImport (898.41s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containerapps	2783.480s
```

https://github.com/iambaim/terraform-provider-azurerm/actions/runs/8861417824/job/24333261609#step:7:1

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New Resource: `azurerm_container_app_environment_custom_domain` [#24346](https://github.com/hashicorp/terraform-provider-azurerm/pull/24346)
* `azurerm_container_app_environment` - expose the `custom_domain_verification_id` attribute [#24346](https://github.com/hashicorp/terraform-provider-azurerm/pull/24346)

This is a (please select all that apply):

- [ ] Bug Fix
- [X] New Feature (ie adding a service, resource, or data source)
- [X] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/24190, fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/20829, fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/20502

### CODE REVIEW

**Comment by iambaim** [comment]
Hi @katbyte, thanks for approving the workflows earlier. I can see one of the checks fails but that seems to be unrelated to my changes.

````
==> Checking for use of gradually deprecated functions...
internal/services/iothub/iothub_resource.go:73:    Create: resourceIotHubCreateUpdate,
New Resources should no longer use combined CreateUpdate methods, please
split these into two separate Create and Update methods.

Existing resources can continue to use combined CreateUpdate methods
for the moment - but over time these will be split into separate Create and
Update methods.
```` 

I have now rebase my commits to the latest main. Hopefully all the checks are good now.

**Comment by iambaim** [comment]
Can I have a review, @jackofallops ? Thanks!

**Comment by davidkarlsen** [comment]
Or maybe @katbyte can review?

**Comment by lgmorand** [comment]
Hi @jackofallops 

anyone who could do the review please ? :) My Cx is looking for the same thing and it blocks the deployment of ACA. I did a code review and it looks fine to me.

ps: @iambaim thanks for the work !

**Comment by goldmark** [comment]
Hi @jackofallops are there any plans to merge this request? any ETA?

**Comment by raddaoui** [comment]
any updates on this PR? my customer is blocked

**Comment by Chigurusetty** [comment]
Hi Team, when this could be available? any ETA please

**Comment by lgmorand** [comment]
@katbyte for visibility. thanks <3

**Comment by iambaim** [comment]
Just synced the code with main. Ready for another round of testing, @katbyte and @jackofallops . Thanks!

**Comment by davidkarlsen** [comment]
@tombuildsstuff can this be merged?

* it's a small thing
* many have voted for it
* it passes
* it is sitting for 4th month on a row waiting for merge
* it's really useful for enterprise customers which have more requirements to networking

**Comment by iambaim** [comment]
Users wanting to implement any form of custom domains in their Azure Container Apps are in a bit of a pickle now as this is still pending and the on-going issue with `azurerm_container_app_custom_domain` that clashes with secrets. (#25196) :(

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @iambaim - Thanks for this PR.

The successful completion of this configuration requires that the `CustomDomainVerificationId` already be present in resolvable DNS so, for this to be supported in Terraform, this will need to be a separate resource, as is necessary for `azurerm_container_app_custom_domain`, perhaps called `azurerm_container_app_environment_custom_domain`. This virtual resource can then update the Environment after the verification ID is known, and a DNS record can be created in the correct sequence to be present (and used as reference to ensure the dependency graph is in the correct order).

The `CustomDomainVerificationId` should also be added to the `azurerm_container_app_environment` Data Source when the resource is updated.

Can you take a look at splitting this into a new virtual resource, and adding the property to the Data Source, and we'll pick up review from there?

Thanks!

**Comment by iambaim** [comment]
Hi @jackofallops , thanks for the suggestion. I agree, this is a better approach. WIll update the branch in a few days :)

File: internal/services/containerapps/container_app_environment_resource.go [hunk 1]
```diff
@@ -136,6 +165,13 @@ func (r ContainerAppEnvironmentResource) Arguments() map[string]*pluginsdk.Schem
 
 func (r ContainerAppEnvironmentResource) Attributes() map[string]*pluginsdk.Schema {
 	return map[string]*pluginsdk.Schema{
+		"custom_domain_verification_id": {
+			Type:        pluginsdk.TypeString,
+			Computed:    true,
+			Sensitive:   true,
```
[the remark this answers is not in the document]
Comment (iambaim) [reply internal/services/containerapps/container_app_environment_resource.go]: Agreed

**Comment by iambaim** [comment]
Hello @jackofallops . The codes are now ready for review.

FYI, the addition of `testdata/testacc_nopassword.pfx` file is necessary to test the resource update feature, because updating using the same cert will trigger an Azure API error. I copied the file from `internal/services/web/testdata/app_service_certificate_nopassword.pfx`.

Thanks!

File: website/docs/r/container_app_environment_custom_domain.html.markdown [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  location                   = azurerm_resource_group.example.location
+  resource_group_name        = azurerm_resource_group.example.name
+  log_analytics_workspace_id = azurerm_log_analytics_workspace.example.id
+}
+
+resource "azurerm_container_app_environment_custom_domain" "example" {
+  container_app_environment_id          = azurerm_container_app_environment.example.id
+  custom_domain_certificate_blob_base64 = filebase64("testacc.pfx")
+  custom_domain_certificate_password    = "TestAcc"
+  custom_domain_dns_suffix              = "acceptancetest.contoso.com"
```
Comment (katbyte) [inline website/docs/r/container_app_environment_custom_domain.html.markdown:40]: as the parent resource is `___custom_domain` i think we can make these:
```suggestion
  certificate_blob_base64 = filebase64("testacc.pfx")
  certificate_password    = "TestAcc"
  dns_suffix              = "acceptancetest.contoso.com"
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/container_app_environment_custom_domain.html.markdown [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+* `container_app_environment_id` - (Required) The ID of the Container Apps Managed Environment. Changing this forces a new resource to be created.
+
+* `certificate_blob_base64` - (Required) The bundle of Private Key and Certificate for the Custom DNS Suffix as a base64 encoded PFX or PEM.
+
+* `certificate_password` - (Required) The password for the Certificate bundle.
+
+* `dns_suffix` - (Required) Custom DNS Suffix for the Container App Environment.
+
+
+
```
Comment (katbyte) [inline website/docs/r/container_app_environment_custom_domain.html.markdown:58]:
```suggestion

```

File: website/docs/r/container_app_environment_custom_domain.html.markdown [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  log_analytics_workspace_id = azurerm_log_analytics_workspace.example.id
+}
+
+resource "azurerm_container_app_environment_custom_domain" "example" {
+  container_app_environment_id = azurerm_container_app_environment.example.id
+  certificate_blob_base64      = filebase64("testacc.pfx")
+  certificate_password         = "TestAcc"
+  dns_suffix                   = "acceptancetest.contoso.com"
+}
+
```
Comment (katbyte) [inline website/docs/r/container_app_environment_custom_domain.html.markdown:42]:
```suggestion
```

**Review by katbyte — APPROVED** [review]
aside from a few minor comments this LGTM 🕯️

**Comment by lgmorand** [comment]
thanks @iambaim & @katbyte

**Pull request merged**